File: adapters/dotnet/tests/Nanocached.Caching.Tests/MockNode.cs
using System.Collections.Concurrent;
using System.Net;
using System.Net.Sockets;
using System.Text;

namespace Nanocached.Caching.Tests;

/// <summary>
/// A minimal in-process nanocached node speaking just the slice of the
/// wire protocol the .NET caching adapter's traffic produces: the
/// <c>A ... T</c> handshake, the namespaced <c>g</c>/<c>s</c>/<c>d</c>/<c>c</c>
/// commands, their legacy default-namespace forms (the SDK's keep-alive
/// probe), and <c>F</c>. Mirrors the structure of
/// adapters/spring/src/test/java/org/nanocached/spring/MockNode.java (a
/// trimmed re-implementation for the same reason that one is: the SDK's own
/// test doubles are internal to the SDK) — not a real store, no TTL expiry,
/// no LRU. Tests only assert what reaches the wire and what comes back.
/// </summary>
internal sealed class MockNode : IDisposable
{
    internal sealed record Entry(byte[] Value, long TtlSeconds);

    private readonly TcpListener _listener;
    private readonly CancellationTokenSource _cts = new();
    private readonly Task _acceptLoop;

    // namespace -> key (both Base64-of-raw-bytes, for value equality) -> entry.
    private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, Entry>> _stores = new();
    private int _clearCount;
    private int _flushCount;
    private int _setCount;

    internal MockNode()
    {
        _listener = new TcpListener(IPAddress.Loopback, 0);
        _listener.Start();
        _acceptLoop = Task.Run(AcceptLoopAsync);
    }

    internal int Port => ((IPEndPoint)_listener.LocalEndpoint).Port;

    /// <summary>How many <c>c</c> (clear-one-namespace) requests this node
    /// has received. Unused by IDistributedCache itself (the SPI has no
    /// Clear), kept for parity with the wire commands this mock
    /// understands.</summary>
    internal int ClearCount => Volatile.Read(ref _clearCount);
    internal int FlushCount => Volatile.Read(ref _flushCount);

    /// <summary>How many <c>s</c>/<c>S</c> (set) requests this node has
    /// received — lets a test prove a re-set actually reached the wire
    /// (e.g. sliding expiration's renew-on-Get), not merely that the
    /// stored value still looks the same.</summary>
    internal int SetCount => Volatile.Read(ref _setCount);

    /// <summary>The entry currently stored for <paramref name="key"/> in
    /// namespace <paramref name="ns"/> (<c>""</c> for the default/legacy
    /// namespace) — what tests assert against to see exactly what reached
    /// the wire, TTL included.</summary>
    internal Entry? EntryFor(string ns, byte[] key) =>
        _stores.TryGetValue(ns, out var store) && store.TryGetValue(EncodeKey(key), out Entry? entry)
            ? entry
            : null;

    private static string EncodeKey(byte[] bytes) => Convert.ToBase64String(bytes);

    private ConcurrentDictionary<string, Entry> Store(string ns) =>
        _stores.GetOrAdd(ns, static _ => new ConcurrentDictionary<string, Entry>());

    public void Dispose()
    {
        _cts.Cancel();
        _listener.Stop();
        _cts.Dispose();
    }

    private async Task AcceptLoopAsync()
    {
        while (!_cts.IsCancellationRequested)
        {
            TcpClient client;
            try
            {
                client = await _listener.AcceptTcpClientAsync(_cts.Token).ConfigureAwait(false);
            }
            catch (Exception) when (_cts.IsCancellationRequested)
            {
                return;
            }
            _ = Task.Run(() => ServeAsync(client));
        }
    }

    private async Task ServeAsync(TcpClient client)
    {
        using TcpClient owned = client;
        NetworkStream stream = owned.GetStream();
        bool tagged = false;
        try
        {
            while (true)
            {
                string[] parts = (await ReadLineAsync(stream).ConfigureAwait(false)).Split(' ');
                string tagSuffix = tagged ? " " + parts[^1] : "";
                switch (parts[0])
                {
                    case "A":
                    {
                        byte[] secret = await ReadExactlyAsync(stream, int.Parse(parts[1])).ConfigureAwait(false);
                        bool accepted = secret.Length > 0;
                        tagged = accepted && parts.Length > 2 && parts[2] == "T";
                        await ReplyAsync(stream, accepted ? (tagged ? "OnT\n" : "On\n") : "En\n")
                            .ConfigureAwait(false);
                        if (!accepted) return;
                        break;
                    }
                    case "G":
                        await GetAsync(stream, "", int.Parse(parts[1]), tagSuffix).ConfigureAwait(false);
                        break;
                    case "g":
                    {
                        byte[] ns = await ReadExactlyAsync(stream, int.Parse(parts[1])).ConfigureAwait(false);
                        await GetAsync(stream, Ns(ns), int.Parse(parts[2]), tagSuffix).ConfigureAwait(false);
                        break;
                    }
                    case "S":
                        await SetAsync(stream, "", parts, 1, tagged, tagSuffix).ConfigureAwait(false);
                        break;
                    case "s":
                    {
                        // Body order is namespace-first, but the length
                        // headers put the namespace length first too, so
                        // read it before delegating for the key/value.
                        byte[] ns = await ReadExactlyAsync(stream, int.Parse(parts[1])).ConfigureAwait(false);
                        await SetAsync(stream, Ns(ns), parts, 2, tagged, tagSuffix).ConfigureAwait(false);
                        break;
                    }
                    case "D":
                        await DeleteAsync(stream, "", int.Parse(parts[1]), tagSuffix).ConfigureAwait(false);
                        break;
                    case "d":
                    {
                        byte[] ns = await ReadExactlyAsync(stream, int.Parse(parts[1])).ConfigureAwait(false);
                        await DeleteAsync(stream, Ns(ns), int.Parse(parts[2]), tagSuffix).ConfigureAwait(false);
                        break;
                    }
                    case "c":
                    {
                        byte[] ns = await ReadExactlyAsync(stream, int.Parse(parts[1])).ConfigureAwait(false);
                        _stores.TryRemove(Ns(ns), out _);
                        Interlocked.Increment(ref _clearCount);
                        await ReplyAsync(stream, "C" + tagSuffix + "\n").ConfigureAwait(false);
                        break;
                    }
                    case "F":
                        _stores.Clear();
                        Interlocked.Increment(ref _flushCount);
                        await ReplyAsync(stream, "C" + tagSuffix + "\n").ConfigureAwait(false);
                        break;
                    default:
                        throw new IOException($"unexpected command {parts[0]}");
                }
            }
        }
        catch (Exception)
        {
            // connection closed by the client (or test teardown)
        }
    }

    private async Task GetAsync(NetworkStream stream, string ns, int keyLength, string tagSuffix)
    {
        byte[] key = await ReadExactlyAsync(stream, keyLength).ConfigureAwait(false);
        Entry? entry = EntryFor(ns, key);
        if (entry is null)
        {
            await ReplyAsync(stream, "N" + tagSuffix + "\n").ConfigureAwait(false);
            return;
        }
        byte[] header = Encoding.ASCII.GetBytes($"V {entry.Value.Length}{tagSuffix}\n");
        await stream.WriteAsync(header).ConfigureAwait(false);
        await stream.WriteAsync(entry.Value).ConfigureAwait(false);
        await stream.FlushAsync().ConfigureAwait(false);
    }

    private async Task SetAsync(
        NetworkStream stream, string ns, string[] parts, int firstLengthIndex, bool tagged, string tagSuffix)
    {
        int keyLength = int.Parse(parts[firstLengthIndex]);
        int valueLength = int.Parse(parts[firstLengthIndex + 1]);
        // Remaining numeric fields: [ttl] in untagged mode, [ttl] tag in
        // tagged mode — the tag is always last.
        int remaining = parts.Length - (firstLengthIndex + 2) - (tagged ? 1 : 0);
        long ttlSeconds = remaining > 0 ? long.Parse(parts[firstLengthIndex + 2]) : 0;
        byte[] key = await ReadExactlyAsync(stream, keyLength).ConfigureAwait(false);
        byte[] value = await ReadExactlyAsync(stream, valueLength).ConfigureAwait(false);
        Store(ns)[EncodeKey(key)] = new Entry(value, ttlSeconds);
        Interlocked.Increment(ref _setCount);
        await ReplyAsync(stream, "S" + tagSuffix + "\n").ConfigureAwait(false);
    }

    private async Task DeleteAsync(NetworkStream stream, string ns, int keyLength, string tagSuffix)
    {
        byte[] key = await ReadExactlyAsync(stream, keyLength).ConfigureAwait(false);
        bool existed = Store(ns).TryRemove(EncodeKey(key), out _);
        await ReplyAsync(stream, (existed ? "D" : "N") + tagSuffix + "\n").ConfigureAwait(false);
    }

    private static string Ns(byte[] namespaceBytes) => Encoding.UTF8.GetString(namespaceBytes);

    private static async Task ReplyAsync(NetworkStream stream, string line)
    {
        await stream.WriteAsync(Encoding.ASCII.GetBytes(line)).ConfigureAwait(false);
        await stream.FlushAsync().ConfigureAwait(false);
    }

    private static async Task<byte[]> ReadExactlyAsync(NetworkStream stream, int length)
    {
        byte[] buffer = new byte[length];
        int offset = 0;
        while (offset < length)
        {
            int read = await stream.ReadAsync(buffer.AsMemory(offset, length - offset)).ConfigureAwait(false);
            if (read == 0) throw new IOException("connection closed");
            offset += read;
        }
        return buffer;
    }

    private static async Task<string> ReadLineAsync(NetworkStream stream)
    {
        var line = new StringBuilder();
        var one = new byte[1];
        while (true)
        {
            int read = await stream.ReadAsync(one.AsMemory(0, 1)).ConfigureAwait(false);
            if (read == 0) throw new IOException("connection closed");
            if (one[0] == (byte)'\n') return line.ToString();
            line.Append((char)one[0]);
        }
    }
}

File: adapters/spring/src/test/java/org/nanocached/spring/MockNode.java
package org.nanocached.spring;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * A minimal in-process nanocached node speaking just the slice of the
 * wire protocol the Spring adapter's client traffic produces: the
 * {@code A ... T} handshake, the namespaced {@code g}/{@code s}/{@code
 * d}/{@code c} commands, their legacy default-namespace forms (the
 * SDK's keep-alive probe), the compare-and-set {@code k} (issue #141)
 * {@link NanocachedCache#putIfAbsent} relies on, and {@code F}. A
 * trimmed re-implementation of the SDK's own test double (that one is
 * package-private to the SDK), not a real store: no TTL expiry, no LRU —
 * the adapter tests only assert what reaches the wire and what comes
 * back.
 */
final class MockNode implements AutoCloseable {

    record Entry(byte[] value, long ttlSeconds) {}

    private final ServerSocket server;
    private final Thread acceptLoop;
    /** namespace (UTF-8-ish, raw bytes wrapped) → key → entry. */
    final Map<ByteBuffer, Map<ByteBuffer, Entry>> stores = new ConcurrentHashMap<>();
    final AtomicInteger clearCount = new AtomicInteger();
    final AtomicInteger flushCount = new AtomicInteger();
    final AtomicInteger casSetCount = new AtomicInteger();

    MockNode() throws IOException {
        server = new ServerSocket(0, 16, java.net.InetAddress.getLoopbackAddress());
        acceptLoop = new Thread(() -> {
            while (true) {
                try {
                    Socket socket = server.accept();
                    Thread serve = new Thread(() -> serve(socket));
                    serve.setDaemon(true);
                    serve.start();
                } catch (IOException stop) {
                    return;
                }
            }
        });
        acceptLoop.setDaemon(true);
        acceptLoop.start();
    }

    int port() {
        return server.getLocalPort();
    }

    Map<ByteBuffer, Entry> store(String namespace) {
        return stores.computeIfAbsent(
                ByteBuffer.wrap(namespace.getBytes(StandardCharsets.UTF_8)),
                unused -> new ConcurrentHashMap<>());
    }

    Entry entry(String namespace, byte[] key) {
        return store(namespace).get(ByteBuffer.wrap(key));
    }

    @Override
    public void close() throws IOException {
        server.close();
    }

    private void serve(Socket socket) {
        try (socket) {
            InputStream in = socket.getInputStream();
            OutputStream out = socket.getOutputStream();
            boolean tagged = false;
            while (true) {
                String[] parts = readLine(in).split(" ");
                String tagSuffix = tagged ? " " + parts[parts.length - 1] : "";
                switch (parts[0]) {
                    case "A" -> {
                        byte[] secret = in.readNBytes(Integer.parseInt(parts[1]));
                        boolean accepted = secret.length > 0;
                        tagged = accepted && parts.length > 2 && parts[2].equals("T");
                        reply(out, (accepted ? (tagged ? "OnT\n" : "On\n") : "En\n"));
                        if (!accepted) {
                            return;
                        }
                    }
                    case "G" -> get(in, out, "", Integer.parseInt(parts[1]), tagSuffix);
                    case "g" -> {
                        byte[] namespace = in.readNBytes(Integer.parseInt(parts[1]));
                        get(in, out, ns(namespace), Integer.parseInt(parts[2]), tagSuffix);
                    }
                    case "S" -> set(in, out, "", parts, 1, tagged, tagSuffix);
                    case "s" -> {
                        // Body order is namespace-first, but the length
                        // headers put the namespace length first too, so
                        // read it before delegating for the key/value.
                        byte[] namespace = in.readNBytes(Integer.parseInt(parts[1]));
                        set(in, out, ns(namespace), parts, 2, tagged, tagSuffix);
                    }
                    case "D" -> delete(in, out, "", Integer.parseInt(parts[1]), tagSuffix);
                    case "d" -> {
                        byte[] namespace = in.readNBytes(Integer.parseInt(parts[1]));
                        delete(in, out, ns(namespace), Integer.parseInt(parts[2]), tagSuffix);
                    }
                    case "c" -> {
                        byte[] namespace = in.readNBytes(Integer.parseInt(parts[1]));
                        stores.remove(ByteBuffer.wrap(namespace));
                        clearCount.incrementAndGet();
                        reply(out, "C" + tagSuffix + "\n");
                    }
                    case "F" -> {
                        stores.clear();
                        flushCount.incrementAndGet();
                        reply(out, "C" + tagSuffix + "\n");
                    }
                    case "k" -> casSet(in, out, parts, tagged, tagSuffix);
                    default -> throw new IOException("unexpected command " + parts[0]);
                }
            }
        } catch (IOException done) {
            // connection closed by the client (or test teardown)
        }
    }

    private void get(InputStream in, OutputStream out, String namespace, int keyLength,
            String tagSuffix) throws IOException {
        byte[] key = in.readNBytes(keyLength);
        Entry entry = entry(namespace, key);
        if (entry == null) {
            reply(out, "N" + tagSuffix + "\n");
            return;
        }
        out.write(("V " + entry.value().length + tagSuffix + "\n")
                .getBytes(StandardCharsets.US_ASCII));
        out.write(entry.value());
        out.flush();
    }

    private void set(InputStream in, OutputStream out, String namespace, String[] parts,
            int firstLengthIndex, boolean tagged, String tagSuffix) throws IOException {
        int keyLength = Integer.parseInt(parts[firstLengthIndex]);
        int valueLength = Integer.parseInt(parts[firstLengthIndex + 1]);
        // Remaining numeric fields: [ttl] in untagged mode, [ttl] tag in
        // tagged mode — the tag is always last.
        int remaining = parts.length - (firstLengthIndex + 2) - (tagged ? 1 : 0);
        long ttlSeconds = remaining > 0 ? Long.parseLong(parts[firstLengthIndex + 2]) : 0;
        byte[] key = in.readNBytes(keyLength);
        byte[] value = in.readNBytes(valueLength);
        store(namespace).put(ByteBuffer.wrap(key), new Entry(value, ttlSeconds));
        reply(out, "S" + tagSuffix + "\n");
    }

    private void delete(InputStream in, OutputStream out, String namespace, int keyLength,
            String tagSuffix) throws IOException {
        byte[] key = in.readNBytes(keyLength);
        boolean existed = store(namespace).remove(ByteBuffer.wrap(key)) != null;
        reply(out, (existed ? "D" : "N") + tagSuffix + "\n");
    }

    /** {@code k <ns-len> <key-len> <val-len> <cond> [ttl] [tag]},
     * docs/protocol.html "k / x" — only {@code <cond> == "A"} (add if
     * absent) is implemented, the only condition {@link
     * NanocachedCache#putIfAbsent} ever sends. The store's per-namespace
     * map is the lock: check-then-put is synchronized on it so two
     * connections racing on the same key really do serialize the way the
     * real server's CAS does, not just "usually" thanks to test timing. */
    private void casSet(InputStream in, OutputStream out, String[] parts, boolean tagged, String tagSuffix)
            throws IOException {
        byte[] namespace = in.readNBytes(Integer.parseInt(parts[1]));
        int keyLength = Integer.parseInt(parts[2]);
        int valueLength = Integer.parseInt(parts[3]);
        String cond = parts[4];
        byte[] key = in.readNBytes(keyLength);
        byte[] value = in.readNBytes(valueLength);
        casSetCount.incrementAndGet();

        if (!"A".equals(cond)) {
            throw new IOException("MockNode: only cond \"A\" is implemented, got " + cond);
        }
        Map<ByteBuffer, Entry> map = store(ns(namespace));
        ByteBuffer keyBuffer = ByteBuffer.wrap(key);
        boolean stored;
        synchronized (map) {
            stored = !map.containsKey(keyBuffer);
            if (stored) {
                int ttlFieldCount = parts.length - 5 - (tagged ? 1 : 0);
                long ttlSeconds = ttlFieldCount > 0 ? Long.parseLong(parts[5]) : 0;
                map.put(keyBuffer, new Entry(value, ttlSeconds));
            }
        }
        reply(out, (stored ? "S" : "N") + tagSuffix + "\n");
    }

    private static String ns(byte[] namespace) {
        return new String(namespace, StandardCharsets.UTF_8);
    }

    private void reply(OutputStream out, String line) throws IOException {
        out.write(line.getBytes(StandardCharsets.US_ASCII));
        out.flush();
    }

    private static String readLine(InputStream in) throws IOException {
        StringBuilder line = new StringBuilder();
        while (true) {
            int b = in.read();
            if (b < 0) {
                throw new IOException("connection closed");
            }
            if (b == '\n') {
                return line.toString();
            }
            line.append((char) b);
        }
    }
}
